src/metabase/lib/equality.cljc

```diff
@@ -95,6 +96,22 @@
   (lib.options/update-options a-ref (fn [options]
                                       (into {} (remove (fn [[k _v]] (qualified-keyword? k))) options))))
 
+(defn- unjoin-field-name
+  "Convert something like
+
+    [:field {} \"join__col\"]
+
+  to
+
+    [:field {:join-alias \"join\"} \"col\"]"
+  [a-ref]
+  (mbql.u.match/replace a-ref
+    [:field
+     (opts :guard (complement :join-alias))
+     (field-name :guard (every-pred string? #(str/includes? % "__"))) ]
+    (let [[join-alias source-name] (str/split field-name #"__" 2)]
+      [:field (assoc opts :join-alias join-alias) source-name])))
```

Comment:
I only added it at the very end to get my PR green to fix one specific test

https://github.com/metabase/metabase/blob/ec33a542a3807c6672399da2d52af2668a8bc749/test/metabase/lib/field_test.cljc#L195-L210

but I agree that it is icky. Maybe I'll try to see if I can revisit https://github.com/metabase/metabase/commit/962931356435514d331930145506f722421fda2a and fix it in a better way now that this new code is present that doesn't require icky hacks

Reply:
The QP is smart enough that it can understand either

```clj
[:field {:join-alias "Products"} 1] ; products.category
```

or 

```clj
[:field {} "Products__Category"] 
```

and is able to guess what 

```clj
[:field {} 1] ; products.category
```

means, but MLv2 currently doesn't know how to guess the last one, and the FE MLv1 code is generating stuff like that which is pretty busted

Reply:
Ok I think I have a better way to fix this. One sec